### PR TITLE
Feature/37 terraform created files in archive

### LINKED ISSUE #37: ZIP file is not recreated if running in a separate stage
When I modify a file in the `source_directory`, I find that the zip file is **not recreated when one of the files in the source directory has changed**. 

It seems we fixed a part of this when fixing #32 with PR #35 by forcing update on the object in the GCS bucket - but the `archive_file` (zip) itself is not being updated, i.e in the end not forcing the update of the Cloud Function.

The upstream issue for failing to update the `archive_file` is here: https://github.com/terraform-providers/terraform-provider-archive/issues/39. It seems it's linked to running `plan` and `apply` in a different context (think a Cloud Build pipeline composed of two configs that effectively 'forgets' about the `.terraform/` folder between `plan` and `apply`).

I see 3 ways we can work-around this in this module (while waiting for an upstream fix): 

- compute the hash of all files in the source directory and force the recomputing of the `archive_file` resource based on if any hash in the `source_directory` has changed; but I'm not certain how to do this for all files in the `source_directory` (for one file, it would be straightforward - for multiple, we'll need a custom shell script to list out all the files in the directory). 

- have a random id appended to the zip filename that would be forced to regenerate when we need it, i.e by passing e.g a `force_update` variable to the module.

- advertise that `plan` and `apply` should be run in the same context, which would make some users struggle.

Any help is appreciated here :)

### PR DESCRIPTION
**Pull request by taylorludwig**
Fixes #37 

Uses an approach similar to described in https://github.com/terraform-providers/terraform-provider-archive/issues/11

Adds a new optional input var `source_dependent_files` that can take a list of `local_file` resources. 

This then forces the `archive` to not be created until all other terraform created files are in the source directory. Prior to this the archive would be created before any terraform created `local_file`s existed. 

With the extra `null_resource` and `null_data_source` workaround it allows terraform to converge (terraform plan/apply wont show changes unless there are actual file changes in `source_directory` or terraform created `local_file`s). The comment above describes more why the workaround is needed. 

A new example and test is created to demonstrate this by simply having the function return back the value of the random created terraform file. 

Once this is in we can go back and properly fix https://github.com/terraform-google-modules/terraform-google-slo/issues/7

### CODE REVIEW

**Comment by ocervell** [comment]
@taylorludwig thank you so much for this and creating a PR in `terraform-google-slo` too ! Will test this out shortly.

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+### Added
+
+- New for `source_dependent_files` var. If used `archive` won't be created until terraform created `local_file`s are finished. [#38]
```
Comment (aaron-lane) [inline CHANGELOG.md]:
```suggestion
- The `source_dependent_files` variable. If used `archive` won't be created until Terraform created `local_file`s are finished. [#38]
```

File: README.md [hunk 2]
```diff
@@ -26,6 +26,26 @@ The
 is a tested reference of how to use the root module with the
 [event-project-log-entry submodule][event-project-log-entry-submodule].
 
+## Terraform Created Source Files
+
+If you have `local_file` terraform resources that need to be included in the function's archive include them in the optional `source_dependent_files`.
```
Comment (aaron-lane) [inline README.md]:
```suggestion
If you have `local_file` Terraform resources that need to be included in the function's archive include them in the optional `source_dependent_files`.
```

File: examples/automatic-labelling-folder/variables.tf [hunk 3]
```diff
@@ -19,7 +19,7 @@ variable "project_id" {
   description = "The ID of the project to which resources will be applied."
 }
 
-variable "folder_id" {
+variable "sub_folder_id" {
```
Comment (aaron-lane) [inline examples/automatic-labelling-folder/variables.tf]: This change should only be made in the test fixture; `folder_id` is the appropriate name for the example.

File: examples/dynamic-files/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Automatic Labelling for folder projects
+
+This example demonstrates how to use the
+[root module][root-module] that will contain source
+code files generated from terraform itself.
```
Comment (aaron-lane) [inline examples/dynamic-files/README.md]:
```suggestion
code files generated from Terraform itself.
```

File: examples/dynamic-files/main.tf [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+terraform {
+  required_version = "~> 0.12.0"
```
Comment (aaron-lane) [inline examples/dynamic-files/main.tf]: This should go in `versions.tf` and it should set the minimum to 0.12.6 to ensure compatibility with `for_each`.

File: examples/dynamic-files/outputs.tf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+output "project_id" {
+  value       = var.project_id
```
Comment (aaron-lane) [inline examples/dynamic-files/outputs.tf]: We should not need to output variables from examples; this is only necessary in test fixtures to expose values to InSpec.

File: variables.tf [hunk 7]
```diff
@@ -73,6 +73,15 @@ variable "source_directory" {
   description = "The pathname of the directory which contains the function source code."
 }
 
+variable "source_dependent_files" {
+  type = list(object({
+    filename = string
+    id       = string
+  }))
+  description = "A list of any terraform created `local_file`s that the module will wait for before creating the archive."
```
Comment (aaron-lane) [inline variables.tf]:
```suggestion
  description = "A list of any Terraform created `local_file`s that the module will wait for before creating the archive."
```

**Comment by ocervell** [comment]
@taylorludwig thanks so much for this

**Pull request merged**